strongloop/loopback-next
# refactor: remove hand-written index files

1. The first commit moves `if (require.main === module)` code from `/index.js` to `src/index.ts`. This is a preparation to allow us to remove `/index.js` file entirely. It leaves the repository in a broken state, but that's fine because the next commit fixes the problem and will be squashed into this one before landing.

   This change has a nice benefit that it enables TypeScript compiler checks for the code building app config, it discovered a possible problem in `+process.env.PORT` line (`+undefined` returns `NaN`).

   I am also slightly changing exports to improve ergonomics - the type `ApplicationOptions` is exported together with the main application class, so that consumers don't need to import the options from `@loopback/context`.

2. In the next step, I am committing the outcome of running an automated script (see https://gist.github.com/bajtos/68672a8229bd03fddad5a4ad1442061f) to remove top-level index files and update `package.json` metadata.

3. In the third commit I am updating project templates in `packages/cli` to match the new style.

4. Finally I am updating documentation, tutorials and README files in `example/*`.

Resolves #2613, resolves  #2609

## Verification

Here are the steps I performed to verify that this change does not break our source-code editing experience in VS Code:

1. Run `npm run clean` to remove all `dist` folders 
2. Verify that VSCode is still able to understand relations between source code artifacts defined in one package (e.g. `context`) and used elsewhere (e.g. `example-todo`), as described in https://github.com/strongloop/loopback-next/blob/master/docs/site/VSCODE.md:

    > - Verify that "Go to definition" works across package boundaries. Find a place
where we are calling `@inject` in `authentication` package, press F12 to go to
the definition of `inject`. VSCode should jump to the `.ts` file in `src` (as
opposed to jumping to a `.d.ts` file in `dist`)
    > - Verify that refactorings like "rename symbol" (`F2`) will change all places
using the renamed entity. Two different scenarios to verify: rename at the place
where the entity is defined, rename at the place where the entity is used. (You
can e.g. rename `inject` to test.)


## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [x] Affected artifact templates in `packages/cli` were updated
- [x] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈
